internal/cmd/headless_repl/src/processes/lsp_invoke.rs

```diff
@@ -0,0 +1,349 @@
+                        text_document: TextDocumentIdentifier {
+                            uri: (Url::parse("file:///foo.flux").unwrap()),
+                        },
+
+                        position: Position {
+                            line: line_num as u32,
+                            character: character as u32,
+                            // character: text.len() as u32,
+                        },
+                    },
+                    work_done_progress_params: Default::default(),
+                    partial_result_params: Default::default(),
+                    context: Default::default(),
+                })?,
+            );
+            let a = serde_json::to_value(req.id(req_id).finish())?;
+            let headed = add_headers(serde_json::to_string(&a)?);
+            Ok(headed)
+        }
+        "importOpen" => {
+            let req: RequestBuilder = jsonrpc::Request::build(DidOpenTextDocument::METHOD).params(
+                serde_json::to_value(DidOpenTextDocumentParams {
+                    text_document: TextDocumentItem {
+                        uri: Url::parse("file:///import.flux").unwrap(),
+                        language_id: "flux".to_string(),
+                        version: version as i32,
+                        text: "".to_string(),
+                    },
+                })?,
+            );
+            let a = serde_json::to_value(req.id(req_id).finish())?;
+            let headed = add_headers(serde_json::to_string(&a)?);
+            Ok(headed)
+        }
+        "importChange" => {
+            let basic_change = vec![TextDocumentContentChangeEvent {
+                range: None,
+                range_length: None,
+                text: text.to_string(),
+            }];
+            let req: RequestBuilder = jsonrpc::Request::build(DidChangeTextDocument::METHOD)
+                .params(serde_json::to_value(DidChangeTextDocumentParams {
+                    text_document: VersionedTextDocumentIdentifier {
+                        uri: (Url::parse("file:///import.flux").unwrap()),
+                        version: version as i32,
+                    },
+                    content_changes: basic_change,
+                })?);
+            let a = serde_json::to_value(req.id(req_id).finish())?;
+            let headed = add_headers(serde_json::to_string(&a)?);
+            Ok(headed)
+        }
+        // "resolve_completion" =>{
+        // // let req= jsonrpc::Request::build(ResolveCompletionItem::METHOD)
+        // //     .params()
+        // }
+        _ => Err(InvalidRequestType),
+    }
+}
+
+pub fn start_lsp() -> Child {
+    //step one: start the process
+    let mut child = Command::new("flux-lsp")
+        // .arg("")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .expect("failure to execute");
+    child
+}
+
+pub fn join_imports(lock: MutexGuard<HashSet<String>>, delim: &str) -> String {
```

Comment:
```suggestion
pub fn join_imports(lock: &HashSet<String>, delim: &str) -> String {
```

You don't need to pass the whole mutexguard in, just pass it by reference instead. (This will change when the mutex is unlocked, but you can add a `drop(mutex_guard)` after calling `join_imports` if that is relevant.)